crates/typst-cli/src/update.rs

```diff
@@ -0,0 +1,311 @@
+    io::copy(&mut buffer.as_slice(), &mut binary_part).map_err(|err| {
+        fs::remove_file(&temp_exe).ok();
+        eco_format!("failed to write typst_update.part: {}", err)
+    })?;
+
+    self_replace::self_replace(&temp_exe)
+        .map_err(|err| {
+            fs::remove_file(&temp_exe).ok();
+            eco_format!("self replace failed: {}", err)
+        })
+        .and_then(|_| {
+            fs::remove_file(&temp_exe)
+                .map_err(|err| eco_format!("failed to delete typst_update.part: {}", err))
+        })
+}
+
+/// Reflects the posibility of self updating.
+#[derive(Clone, Copy, Debug)]
+enum SelfUpdatePermission {
+    Deny,
+    Permit,
+}
+
+/// Assets belonging to a GitHub release.
+///
+/// Primarily used to download pre-compiled Typst CLI binaries.
+#[derive(Debug, Deserialize)]
+struct Asset {
+    pub name: String,
+    pub browser_download_url: String,
+}
+
+/// A GitHub release.
+#[derive(Debug, Deserialize)]
+struct Release {
+    tag_name: String,
+    assets: Vec<Asset>,
+}
+
+impl Release {
+    /// Download the lastest release from the Typst repository.
+    pub fn from_latest() -> StrResult<Self> {
+        let url = format!(
+            "https://api.github.com/repos/{}/{}/releases/latest",
+            TYPST_GITHUB_ORG, TYPST_REPO
+        );
+
+        Release::download(&url)
+    }
+
+    /// Download the target release from the Typst repository.
+    pub fn from_tag(tag: Version) -> StrResult<Release> {
+        let url = format!(
+            "https://api.github.com/repos/{}/{}/releases/tags/v{}",
+            TYPST_GITHUB_ORG, TYPST_REPO, tag
+        );
+
+        Release::download(&url)
+    }
+
+    fn download(url: &str) -> StrResult<Self> {
+        match ureq::get(url).call() {
+            Ok(response) => {
+                Ok(response.into_json().expect("unable to get json from response"))
+            }
+            Err(ureq::Error::Status(404, _)) => {
+                bail!("release not found (searched at {url})")
+            }
+            Err(_) => bail!("failed to download release (network failed)"),
+        }
+    }
+
+    /// Sorts through the assets from a given [`Release`] and picks the right one
+    /// for the target platform, returning its packed binary.
+    pub fn download_asset(&self, asset_name: &str) -> StrResult<PackedBinary> {
+        let asset = self
+            .assets
+            .iter()
+            .find(|a| a.name.starts_with(asset_name))
+            .ok_or("could not find release for your target platform")?;
+
+        let response = match ureq::get(&asset.browser_download_url).call() {
+            Ok(response) => response,
+            Err(ureq::Error::Status(404, _)) => {
+                bail!("asset not found (searched for {})", asset.name);
+            }
+            Err(_) => bail!("failed to load asset (network failed)"),
+        };
+
+        let mut buffer = response
+            .header("Content-Length")
```

Comment:
> And using Unpack would give you the freedom to relatively easily add another archive type if you ever wanted to support more/different ones.

Adding a third archive type would also be simple in my design, it would just be a function rather than an impl. unless I'm missing something, I still think that it would be simpler. Here's an example of how I would write it:

```rust
    /// Sorts through the assets from a given [`Release`] and picks the right one
    /// for the target platform, downloadings it and extracts the Typst binary.
    pub fn download_binary(&self, asset_name: &str) -> StrResult<Vec<u8>> {
        let asset = self
            .assets
            .iter()
            .find(|a| a.name.starts_with(asset_name))
            .ok_or("could not find release for your target platform")?;

        let response = match ureq::get(&asset.browser_download_url).call() {
            Ok(response) => response,
            Err(ureq::Error::Status(404, _)) => {
                bail!("asset not found (searched for {})", asset.name);
            }
            Err(_) => bail!("failed to load asset (network failed)"),
        };

        let mut data = Vec::new();
        response
            .into_reader()
            .read_to_end(&mut data)
            .map_err(|err| eco_format!("failed to read response buffer: {err}"))?;

        if asset_name.contains("windows") {
            extract_binary_from_zip(&data, asset_name)
        } else {
            extract_binary_from_tar_xz(&data)
        }
    }
}

/// Extract the Typst binary from a ZIP archive.
fn extract_binary_from_zip(data: &[u8], asset_name: &str) -> StrResult<Vec<u8>> {
    let mut archive = ZipArchive::new(Cursor::new(data))
        .map_err(|err| eco_format!("failed to extract ZIP archive: {err}"))?;

    let mut file = archive
        .by_name(&format!("{asset_name}/typst.exe"))
        .map_err(|_| "ZIP archive did not contain Typst binary")?;

    let mut buffer = vec![];
    file.read_to_end(&mut buffer).map_err(|err| {
        eco_format!("failed to read binary data from ZIP archive: {err}")
    })?;

    Ok(buffer)
}

/// Extract the Typst binary from a `.tar.xz` archive.
fn extract_binary_from_tar_xz(data: &[u8]) -> StrResult<Vec<u8>> {
    let mut archive = tar::Archive::new(XzDecoder::new(Cursor::new(data)));

    let mut file = archive
        .entries()
        .map_err(|err| eco_format!("failed to extract tar.xz archive: {err}"))?
        .filter_map(Result::ok)
        .find(|e| e.path().unwrap_or_default().ends_with("typst"))
        .ok_or("tar.xz archive did not contain Typst binary")?;

    let mut buffer = vec![];
    file.read_to_end(&mut buffer).map_err(|err| {
        eco_format!("failed to read binary data from tar.xz archive: {err}")
    })?;

    Ok(buffer)
}
```

Reply:
I think we can skip all `Vec::with_capacity`. This happens just once and is not at all on a hot path. Allocation costs will be far outweighed by network delay. `BufReader` is also not necessary on top of `io::Cursor`.

Reply:
> `BufReader` is also not necessary on top of `io::Cursor`.

Cool, I didn't know that.